README.md
# Rest API

#### _Framework: Express.js_

merupakan rest API yang dibuat untuk memenuhi proses assignment test di Ruang Talenta.

## Routes

Berikut ini merupakan list dari semua routes yang dibuat.
rout terbagi menjadi beberapa bagian sesuai applikasi, diantaranya:

#### 1. Program

| Method | Routes                                   | Description                                                                                  | Example                                          |
| ------ | ---------------------------------------- | -------------------------------------------------------------------------------------------- | ------------------------------------------------ |
| GET    | /program                                 | mendapatkan semua data dari program                                                          | http://localhost:8080/program                    |
| GET    | /program/:id_program                     | mendapatkan data detail dari program yang dipilih                                            | http://localhost:8080/program/2                  |
| GET    | /program/categories                      | mendapatkan data kategori yang ada di program                                                | http://localhost:8080/categories                 |
| GET    | /program/category/:category_name         | mendapatkan data program yang termasuk kedalam kategori pilihan kategori yang ada di program | http://localhost:8080/program/category/kesehatan |
| GET    | /program/search/judul=:cari_nama_program | melakukan pencarian data program berdasarkan judul                                           | http://localhost:8080/program/search/judul=puasa |

#### 2. Berita

| Method | Routes             | Description                                      | Example                        |
| ------ | ------------------ | ------------------------------------------------ | ------------------------------ |
| GET    | /berita            | mendapatkan semua data berita                    | http://localhost:8080/berita   |
| GET    | /berita/:id_berita | mendapatkan data detail dari berita yang dipilih | http://localhost:8080/berita/2 |

#### 3. Donatur

| Method | Routes               | Description                                                           | Example                                                                                                                           |
| ------ | -------------------- | --------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------- |
| GET    | /donatur             | mendapatkan semua data donatur yang terdaftar                         | http://localhost:8080/donatur                                                                                                     |
| GET    | /donatur/:id_donatur | mendapatkan data detail dari donatur yang dipilih                     | http://localhost:8080/donatur/2                                                                                                   |
| POST   | /donatur/add         | melakukan insert data donatur baru. data yang dikirim berformat json. | http://localhost:8080/donatur/add Data yang di kirim adalah : {"nama_lengkap": "Nama lengkap","hp_email": "NomorHp/alamat email"} |

#### 4. Donasi

| Method | Routes                              | Description                                                          | Example                                                                                                                                        |
| ------ | ----------------------------------- | -------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------- |
| GET    | /donasi                             | mendapatkan semua data donasi yang telah dilakukan                   | http://localhost:8080/donasi                                                                                                                   |
| GET    | /donasi/:id_user                    | mendapatkan data detail donasi dari donatur yang dipilih             | http://localhost:8080/donasi/1                                                                                                                 |
| POST   | /donasi/add/:id_program/:id_donatur | melakukan insert data donasi baru. data yang dikirim berformat json. | http://localhost:8080/donasi/add/2/1 Data yang di kirim adalah : {"visibility": "sembunyikan","total_donasi": 50397,"id_metode_pembayaran": 1} |

#### 5. Metode Pembayaran

| Method | Routes                               | Description                                                                                     | Example                                                                                                                                   |
| ------ | ------------------------------------ | ----------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------- |
| GET    | /metode_pembayaran                   | mendapatkan semua data metode pembayaran yang tersia                                            | http://localhost:8080/metode_pembayaran                                                                                                   |
| GET    | /metode_pembayaran/:id_metode        | mendapatkan data detail metode pembayaran yang dipilih                                          | http://localhost:8080/metode_pembayaran/1                                                                                                 |
| POST   | /metode_pembayaran/add               | melakukan insert data metode pembayaran baru. data yang dikirim berformat json.                 | http://localhost:8080/metode_pembayaran/add Data yang di kirim adalah : {"nama_metode_pembayaran":"Dana","nomor_rekening": "121212"}      |
| PUT    | /metode_pembayaran/update/:id_metode | melakukan perubahan data pada metode pembayaran yang dipilih. data yang dikirim berformat json. | http://localhost:8080/metode_pembayaran/update/1 Data yang di kirim adalah : {"nama_metode_pembayaran":"Dana","nomor_rekening": "121212"} |
| DELETE | /metode_pembayaran/delete/:id_metode | melakukan penghapusan data pada metode pembayaran yang dipilih.                                 | http://localhost:8080/metode_pembayaran/delete/1                                                                                          |


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.1.1](https://github.com/gaalih/restExpress/compare/v1.1.0...v1.1.1) (2022-04-26)

## 1.1.0 (2022-04-26)


### Features

* add standard release ([967c180](https://github.com/gaalih/restExpress/commit/967c180e7129825ea7b24a867aff595ffb70d94e))


modules/donasi.js
const express = require('express')
const router = express.Router()
var data_donasi = require('../datas/donasi.json')

// middleware that is specific to this router
router.use((req, res, next) => {
  console.log('Time: ', Date.now())
  next()
})

// all data
router.get('/', (req, res) => {
  res.status(200).send(data_donasi)
})

// specific data from specific user
router.get('/:id_user', (req, res) => {
  const donasi_dicari = data_donasi.find(
    (x) => x.id_donatur === parseInt(req.params.id_user),
  )

  // jika data tidak ditemukan
  if (donasi_dicari == null) {
    res.status(418).send({ message: 'Data Donasi Tidak Ditemukan!' })
  }
  res.send(donasi_dicari)
})

// insert donation
router.post('/add/:id_program/:id_donatur', (req, res) => {
  const { id_program } = req.params
  const { id_donatur } = req.params
  const data_donasi = req.body

  let data_disimpan = {
    message: 'Data berhasil disimpan',
    data: {
      id_donatur: id_program,
      visibility: data_donasi['visibility'],
      id_donasi_program: id_program,
      kode_donasi: Date.now(),
      total_donasi: data_donasi['total_donasi'],
      id_metode_pembayaran: data_donasi['id_metode_pembayaran'],
      tgl_donasi: Date(),
      status_pembayaran: 'menunggu',
    },
  }

  // jika data tidak lengkap
  if (!data_donasi || !id_donatur || !id_program) {
    res.status(418).send({ message: 'Data Yang Anda Kirimkan Tidak Lengkap.' })
  }
  res.status(200).send(data_disimpan)
})

module.exports = router


modules/donatur.js
const express = require('express')
const router = express.Router()
var data_donatur = require('../datas/donatur.json')

// middleware that is specific to this router
router.use((req, res, next) => {
  console.log('Time: ', Date.now())
  next()
})

// all data
router.get('/', (req, res) => {
  res.status(200).send(data_donatur)
})

// specific data
router.get('/:id_donatur', (req, res) => {
  let donatur_dicari = data_donatur.find(
    (x) => x.id_donatur === parseInt(req.params.id_donatur),
  )

  // jika data tidak ditemukan
  if (donatur_dicari == null) {
    res.status(418).send({ message: 'Data Donatur Tidak Ditemukan!' })
  }
  res.send(donatur_dicari)
})

// insert donatur
router.post('/add', (req, res) => {
  const data_donatur = req.body

  var data_disimpan = {
    message: 'Data berhasil disimpan',
    data: {
      id_donatur: Date.now(),
      nama_lengkap: data_donatur['nama_lengkap'],
      hp_email: data_donatur['hp_email'],
    },
  }

  // jika data tidak lengkap
  if (!data_donatur) {
    res.status(418).send({ message: 'Data Yang Anda Kirimkan Tidak Lengkap.' })
  }
  res.status(200).send(data_disimpan)
})

module.exports = router


modules/program.js
const express = require('express')
const router = express.Router()
var data_program = require('../datas/program.json')

// middleware that is specific to this router
router.use((req, res, next) => {
  console.log('Access Date: ', Date.now())
  next()
})

// all data
router.get('/', (req, res) => {
  res.status(200).send(data_program)
})

// all categories
router.get('/categories', (req, res) => {
  let results = []
  for (let i = 0; i < data_program.length; i++) {
    results.push(data_program[i]['kategori'])
  }
  results = [...new Set(results)]
  res.status(200).send(results)
})

// search from specific category
router.get('/category/:category_name', (req, res) => {
  const { category_name } = req.params
  let results = []
  for (let i = 0; i < data_program.length; i++) {
    if (data_program[i]['kategori'] == category_name) {
      results.push(data_program[i])
    }
  }

  if (results.length === 0 || category_name == null) {
    res.status(418).send({
      message: `Data Dengan Kategori ${category_name} Tidak Ditemukan!`,
    })
  }
  res.status(200).send(results)
})

// specific data from id
router.get('/:id_program', (req, res) => {
  const program_dipilih = data_program.find(
    (x) => x.id === parseInt(req.params.id_program),
  )
  // jika data tidak ditemukan
  if (program_dipilih == null) {
    res.status(418).send({ message: 'Data Program Tidak Ditemukan!' })
  }
  res.status(200).send(program_dipilih)
})

// search some data from judul
router.get('/search/judul=:cari_nama_program', (req, res) => {
  const { cari_nama_program } = req.params
  let results = []
  for (let i = 0; i < data_program.length; i++) {
    let judul = data_program[i]['judul']
    if (judul.includes(`${cari_nama_program}`)) {
      results.push(data_program[i])
    }
  }
  // jika data tidak ditemukan
  if (results.length === 0 || cari_nama_program == null) {
    res
      .status(418)
      .send({ message: `Data Program ${cari_nama_program} Tidak Ditemukan!` })
  }
  res.status(200).send(results)
})

module.exports = router


modules/metode_pembayaran.js
const express = require('express')
const router = express.Router()
var metode_pembayaran = require('../datas/metode_pembayaran.json')

// middleware that is specific to this router
router.use((req, res, next) => {
  console.log('Time: ', Date.now())
  next()
})

// all data
router.get('/', (req, res) => {
  res.status(200).send(metode_pembayaran)
})

// specific data
router.get('/:id_metode', (req, res) => {
  let metode_dipilih = metode_pembayaran.find(
    (x) => x.id_metode_pembayaran === parseInt(req.params.id_metode),
  )

  // jika data tidak ditemukan
  if (metode_dipilih == null) {
    res.status(418).send({ message: 'Metode Pembayaran Tidak Ditemukan!' })
  }
  res.send(metode_dipilih)
})

// insert metode pembayaran
router.post('/add', (req, res) => {
  const data_metode = req.body

  let data_disimpan = {
    message: 'Data berhasil disimpan',
    data: {
      id_metode_pembayaran: Date.now(),
      nama_metode_pembayaran: data_metode['nama_metode_pembayaran'],
      nomor_rekening: data_metode['nomor_rekening'],
    },
  }

  // jika data tidak lengkap
  if (!data_metode) {
    res.status(418).send({ message: 'Data Yang Anda Kirimkan Tidak Lengkap.' })
  }
  res.status(200).send(data_disimpan)
})

// update metode pembayaran
router.put('/update/:id_metode', (req, res) => {
  const { id_metode } = req.params
  const data_metode = req.body

  let data_disimpan = {
    message: `Data: ${id_metode} berhasil diubah`,
    data: {
      nama_metode_pembayaran: data_metode['nama_metode_pembayaran'],
      nomor_rekening: data_metode['nomor_rekening'],
    },
  }

  // jika data tidak lengkap
  if (!data_metode) {
    res.status(418).send({ message: 'Data Yang Anda Kirimkan Tidak Lengkap.' })
  }
  res.status(200).send(data_disimpan)
})

router.delete('/delete/:id_metode', (req, res) => {
  const { id_metode } = req.params

  // jika data tidak lengkap
  if (!id_metode) {
    res.status(418).send({ message: 'Data Yang Anda Cari Tidak Tersidia.' })
  }
  res.status(200).send({ message: `Data: ${id_metode} berhasil dihapus` })
})

module.exports = router
